README.md
<p align="center">
<a href="https://github.com/spapanik/mathlib/actions/workflows/build.yml"><img alt="Build" src="https://github.com/spapanik/mathlib/actions/workflows/build.yml/badge.svg"></a>
<a href="https://lgtm.com/projects/g/spapanik/mathlib/alerts/"><img alt="Total alerts" src="https://img.shields.io/lgtm/alerts/g/spapanik/mathlib.svg"/></a>
<a href="https://github.com/spapanik/mathlib/blob/master/LICENSE.txt"><img alt="License" src="https://img.shields.io/github/license/spapanik/mathlib"></a>
<a href="https://pypi.org/project/mathlib"><img alt="PyPI" src="https://img.shields.io/pypi/v/mathlib"></a>
<a href="https://pepy.tech/project/mathlib"><img alt="Downloads" src="https://pepy.tech/badge/mathlib"></a>
<a href="https://github.com/psf/black"><img alt="Code style" src="https://img.shields.io/badge/code%20style-black-000000.svg"></a>
</p>

# mathlib

_mathlib_ is a pure python mathematics library.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog][clog], and this project adheres to [Semantic Versioning][semver].

## [Unreleased]

### Added
- Added python 3.10 support
- Added a method to calculate the integer square root of a number

## [0.2.0] - 2020-10-13
### Changed
- Improved the lcm to take an arbitrary number of arguments
- Improved the modular inverse to use the stdlib one

### Added
- Added a method to calculate polygonal numbers
- Added a method to calculate gcd
- Added a method to calculate the product of an interable

## [0.1.4] - 2020-09-18
### Added
- Added more type hints

## [0.1.3] - 2019-04-30
### Added
- Made the initial values in method to return the nth Fibonacci number configurable
- Added a method to find the modular multiplicative inverse of a number
- Added a method to calculate n choose k
- Added a method to calculate σ_x(n) (the divisor function)

## [0.1.2] - 2019-04-29
### Added
- Added a method to return the nth Fibonacci number
- Added a generator of primes up to a bound
- Added a generator of Fibonacci numbers
- Added a method to return the least common multiple of two numbers

## [0.1.1] - 2019-04-26
### Added
- Added a method to give the next prime
- Added a generator of prime numbers

## [0.1.0] - 2019-04-18
### Added
- Added a method to check whether a number is prime


[unreleased]: https://github.com/spapanik/mathlib/compare/v0.2.0...master
[0.2.0]: https://github.com/spapanik/mathlib/compare/v0.1.4...v0.2.0
[0.1.4]: https://github.com/spapanik/mathlib/compare/v0.1.3...v0.1.4
[0.1.3]: https://github.com/spapanik/mathlib/compare/v0.1.2...v0.1.3
[0.1.2]: https://github.com/spapanik/mathlib/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/spapanik/mathlib/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/spapanik/mathlib/releases/tag/v0.1.0

[clog]: https://keepachangelog.com/en/1.0.0/
[semver]: https://semver.org/spec/v2.0.0.html


tests/mathlib/test_primes.py
import pytest

from mathlib import primes


@pytest.mark.parametrize(
    ["upper_bound", "prime_numbers"],
    [
        [2, []],
        [3, [2]],
        [4, [2, 3]],
        [5, [2, 3]],
        [6, [2, 3, 5]],
        [11, [2, 3, 5, 7]],
        [12, [2, 3, 5, 7, 11]],
    ],
)
def test_sieve_two_arguments(upper_bound, prime_numbers):
    assert list(primes.sieve(upper_bound)) == prime_numbers


@pytest.mark.parametrize(
    "number", [2, 3, 5, 41, 97, 1234577, 9000011, 21326017, 593441861]
)
def test_is_prime(number):
    assert primes.is_prime(number) is True


@pytest.mark.parametrize(
    "number",
    [
        1,
        4,
        6,
        25,
        2133,
        1373889,
        9080339,
        25326215,
        4759123293,
        1122004669803,
        2152302898947,
        3474749660601,
        341550071728525,
        3825123056546413239,
        318665857834031151167663,
        3317044064679887385962127,
    ],
)
def test_is_not_prime(number):
    assert primes.is_prime(number) is False


@pytest.mark.parametrize(
    ["number", "prime"], [[1, 2], [2, 3], [3, 5], [91, 97], [1000, 1009]]
)
def test_next_prime(number, prime):
    assert primes.next_prime(number) == prime


def test_primes():
    prime_numbers = []
    for prime in primes.primes():
        if prime > 30:
            break
        prime_numbers.append(prime)
    assert prime_numbers == [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]


@pytest.mark.parametrize(
    ["n", "divisors"],
    [[10, [1, 2, 5, 10]], [12, [1, 2, 3, 4, 6, 12]], [13, [1, 13]]],
)
def test_divisor_sigma(n, divisors):
    for x in range(5):
        expected = sum(map(lambda d: d ** x, divisors))
        assert primes.divisor_sigma(n, x) == expected


tests/mathlib/test_numbers.py
import pytest

from mathlib import numbers


@pytest.mark.parametrize(
    ["integers", "expected"],
    [
        [[], 0],
        [[45], 45],
        [[0, 8], 8],
        [[1, 2], 1],
        [[6, 4], 2],
        [[4, 2, 198, 6], 2],
    ],
)
def test_gcd(integers, expected):
    assert numbers.gcd(*integers) == expected


@pytest.mark.parametrize(
    ["integers", "expected"],
    [
        [[], 1],
        [[45], 45],
        [[0, 8], 0],
        [[1, 2], 2],
        [[6, 4], 12],
        [[1, 2, 3, 4, 6], 12],
    ],
)
def test_lcm(integers, expected):
    assert numbers.lcm(*integers) == expected


@pytest.mark.parametrize(["a", "b"], [[7, 3], [872, 7959], [7959, 872], [42, 35129]])
def test_modular_inverse(a, b):
    inverse = numbers.modular_inverse(a, b)
    assert 0 < inverse < b
    assert (a * inverse) % b == 1


@pytest.mark.parametrize(
    ["n", "expected"],
    [[0, 1], [1, 1], [2, 2], [-11, 144], [-10, -89], [25, 121393], [40, 165580141]],
)
def test_fibonacci(n, expected):
    assert numbers.fibonacci(n) == expected


@pytest.mark.parametrize(
    ["initial", "count", "expected"],
    [[[0, 7], 4, [0, 7, 7, 14]], [[1, 1], 5, [1, 1, 2, 3, 5]]],
)
def test_fibonacci_numbers(initial, count, expected):
    fibonacci = numbers.fibonacci_numbers(*initial)
    result = [next(fibonacci) for _ in range(count)]
    assert result == expected


@pytest.mark.parametrize(
    ["n", "k", "expected"],
    [
        [0, 0, 1],
        [7, 0, 1],
        [7, 1, 7],
        [7, 2, 21],
        [7, 3, 35],
        [7, 4, 35],
        [7, 5, 21],
        [7, 6, 7],
        [7, 7, 1],
        [8, 0, 1],
        [8, 1, 8],
        [8, 2, 28],
        [8, 3, 56],
        [8, 4, 70],
        [8, 5, 56],
        [8, 6, 28],
        [8, 7, 8],
        [8, 8, 1],
    ],
)
def test_binomial(n, k, expected):
    assert numbers.binomial(n, k) == expected


@pytest.mark.parametrize(
    ["s", "n", "expected"],
    [
        [3, 4, 10],
        [4, 8, 64],
        [5, 5, 35],
        [6, 3, 15],
    ],
)
def test_polygonal_number(s, n, expected):
    assert numbers.polygonal_number(s, n) == expected


src/mathlib/_seven.py
import math
import operator
import sys
from functools import reduce

MINOR = sys.version_info.minor


if MINOR >= 8:

    def modular_inverse(n: int, mod: int) -> int:
        return pow(n, -1, mod)

    prod = math.prod
    isqrt = math.isqrt
else:

    def _integer_sqrt(x: int, n: int) -> int:
        return (x + n // x) >> 1

    def isqrt(n: int) -> int:
        if n < 0:
            raise ValueError("isqrt() argument must be nonnegative")

        if n == 0:
            return 0

        initial_guess, next_guess = n, _integer_sqrt(n, n)
        while initial_guess > next_guess:
            initial_guess, next_guess = next_guess, _integer_sqrt(next_guess, n)
        return next_guess

    def prod(iterable, *, start=1):
        return reduce(operator.mul, iterable, start)

    def modular_inverse(n: int, mod: int) -> int:
        original_modulo = mod
        x = 1
        y = 0

        while n > 1:
            x, y = y, x - (n // mod) * y
            n, mod = mod, n % mod

        if x < 0:
            x += original_modulo

        return x


if MINOR >= 9:
    gcd = math.gcd
    lcm = math.lcm
else:

    def gcd(*integers: int) -> int:
        return reduce(math.gcd, integers, 0)

    def lcm(*integers: int) -> int:
        def _lcm(a, b):
            return a * b // gcd(a, b)

        return reduce(_lcm, integers, 1)


src/mathlib/numbers.py
from typing import Iterator

from mathlib import _seven


def gcd(*integers: int) -> int:
    """
    Find the greatest common divisor of the arguments.

    The greatest common divisor of a and 0 is always a, as this
    coincides with gcd to be the greatest lower bound in the
    lattice of divisibility.
    """
    return _seven.gcd(*integers)


def lcm(*integers: int) -> int:
    """
    Find the least common multiple of the arguments.

    The least common multiple of a and 0 is always 0, as this
    coincides with lcm to be the least upper bound in the
    lattice of divisibility.
    """
    return _seven.lcm(*integers)


def isqrt(n: int) -> int:
    """Find the largest integer whose square is less than n."""
    return _seven.isqrt(n)


def modular_inverse(n: int, mod: int) -> int:
    """
    Find the modular inverse of n modulo mod.

    In python 3.6 and 3.7, the algorithm is based on using
    the Extended Euclid Algorithm, to solve the diophantine
    equation n*x + mod*y = 1.

    In later versions this is just a wrapper over the pow function.
    """
    return _seven.modular_inverse(n, mod)


def fibonacci(n: int, a: int = 1, b: int = 1) -> int:
    """
    Return the nth Fibonacci number.

    n can be a negative integer as well.
    """
    values = {0: a, 1: b}

    def fib(m: int) -> int:
        if m in values:
            return values[m]

        k = m // 2
        if m & 1 == 1:
            value = fib(k) * (fib(k + 1) + fib(k - 1))
        else:
            value = fib(k) * fib(k) + fib(k - 1) * fib(k - 1)
        values[m] = value
        return value

    if n < 0:
        if n % 2 == 0:
            return -fib(-n)
        return fib(-n)
    return fib(n)


def fibonacci_numbers(a: int = 0, b: int = 1) -> Iterator[int]:
    """
    Make an iterator that returns the Fibonacci numbers.

    The Fibonacci sequence is configurable, in the sense that the two
    initial values of it can be passed as arguments.
    """
    while True:
        yield a
        a, b = b, a + b


def binomial(n: int, k: int) -> int:
    """
    Calculate n choose k.

    Calculation is using the multiplicative formula, and is performed
    from the side that will minimise the number of calculations.
    """
    output = 1
    k = min(k, n - k)
    for t in range(k):
        output = (n - t) * output // (t + 1)

    return output


def polygonal_number(s: int, n: int) -> int:
    """
    Calculate the n-th s-gonal number.
    """

    return (s - 2) * n * (n - 1) // 2 + n


src/mathlib/primes.py
import math
from itertools import chain, count
from typing import Iterable, Iterator


class UnreachableError(Exception):
    """
    Make mypy happy; this part of the code is unreachable
    """


def sieve(upper_bound: int) -> Iterator[int]:
    """
    Make an iterator that returns the primes up to upper_bound

    This method uses the sieve of Eratosthenes to return the
    primes.
    """
    if upper_bound <= 5:
        if upper_bound > 2:
            yield 2

        if upper_bound > 3:
            yield 3

        return

    yield 2

    upper_bound = (upper_bound >> 1) - 1
    indices = [True] * upper_bound
    end_range = int(math.sqrt(upper_bound)) + 1
    for i in range(end_range):
        if indices[i]:
            slice_start = 2 * i * i + 6 * i + 3
            slice_step = 2 * i + 3
            number_of_primes = math.ceil((upper_bound - slice_start) / slice_step)
            indices[slice_start::slice_step] = [False] * number_of_primes

    for i in range(upper_bound):
        if indices[i]:
            yield 2 * i + 3


def _miller_rabin_loop(witness: int, mantissa: int, power: int, n: int) -> bool:
    if pow(witness, mantissa, n) == 1:
        return False

    for r in range(power):
        if pow(witness, mantissa * (1 << r), n) + 1 == n:
            return False

    return True


def _miller_rabin_witnesses(n: int) -> Iterable[int]:
    if n < 2047:
        return (2,)

    if n < 1373653:
        return 2, 3

    if n < 9080191:
        return 31, 73

    if n < 25326001:
        return 2, 3, 5

    if n < 4759123141:
        return 2, 7, 61

    if n < 1122004669633:
        return 2, 13, 23, 1662803

    if n < 2152302898747:
        return 2, 3, 5, 7, 11

    if n < 3474749660383:
        return 2, 3, 5, 7, 11, 13

    if n < 341550071728321:
        return 2, 3, 5, 7, 11, 13, 17

    if n < 3825123056546413051:
        return 2, 3, 5, 7, 11, 13, 17, 19, 23

    if n < 318665857834031151167461:
        return 2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37

    if n < 3317044064679887385961981:
        return 2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41

    return range(2, math.floor(2 * (math.log(n) ** 2)) + 1)


def is_prime(n: int) -> bool:
    """
    Check if n is a prime number.

    This is a deterministic primality test, but it relies on GHR. This
    seems a good enough compromise. It is very fast for up to 81-bit
    integers, after which it is starts slowing down, due to the fact
    that we need to check for all possible Miller-Rabin witnesses.
    """
    if n < 5:
        return n in (2, 3)

    if n % 2 == 0:
        return False

    mantissa, power = n - 1, 0
    while mantissa & 1 == 0:
        mantissa >>= 1
        power += 1

    for witness in _miller_rabin_witnesses(n):
        if _miller_rabin_loop(witness, mantissa, power, n):
            return False

    return True


def next_prime(n: int) -> int:
    """
    Get the smallest prime that is larger than n.
    """
    if n < 2:
        return 2

    if n == 2:
        return 3

    if n & 1 == 1:
        n += 2
    else:
        n += 1

    for n in count(n, 2):
        if is_prime(n):
            return n

    raise UnreachableError("A prime will be reached")


def primes() -> Iterator[int]:
    """
    Make an iterator that returns the prime numbers in ascending order.
    """
    yield 2

    for n in count(3, 2):
        if is_prime(n):
            yield n


def divisor_sigma(n: int, x: int = 0) -> int:
    """
    Calculate the sum of the xth powers of the positive divisors of n
    """
    out = 1
    for prime_div in chain([2], count(3, 2)):
        power = 1
        while n % prime_div == 0:
            power += 1
            n //= prime_div
        if power != 1:
            if x == 0:
                out *= power
            elif x == 1:
                out *= (prime_div ** power - 1) // (prime_div - 1)
            else:
                out *= (prime_div ** (x * power) - 1) // (prime_div ** x - 1)
            if n == 1:
                return out

    raise UnreachableError("At some point divisors will be exhausted")
